logging/WFCORE-4807-log-routing.adoc

```diff
@@ -0,0 +1,112 @@
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently any modules that log generally log onto the system log context. In some cases, for example a deployment using
+a logging profile, the loggers may end up getting created on the wrong `LogContext`. As an example if `a.war` and `b.war`
+both depend on module `org.jboss.custom.module` and at least one deployment uses a logging profile whichever deployment
+initializes the custom module first will have the loggers created on that `LogContext`. If a logging profile is being
+used it would be that profiles log context. This means if `a.war` uses a logging profile and the module is accessed first
+by `a.war`, then static loggers will be created on the logging profiles `LogContext`. Once `b.war` invokes something to
+be logged in the custom module, then `b.war` 's log messages end up in the logging profiles `LogContext` instead of the
+systems `LogContext`.
+
+https://issues.jboss.org/browse/LOGMR-264[LOGMR-264] will introduce a way to route log messages based on the log context.
+This would allow messages from modules to be routed to deployment specific log contexts. It would also introduce a
+change in behavior which is the reason for an RFE. A new attribute, `allow-log-routing` will be introduced to disable
+the new routing and allow the old behavior.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4897[WFCORE-4807]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1420[EAP7-1420]
+* https://issues.jboss.org/browse/LOGMR-264[LOGMR-264]
+* https://issues.redhat.com/browse/WFCORE-3376[WFCORE-3376]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mkopecky@redhat.com[Marek Kopeck&#253;]
+
+=== Testing By
+
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* WildFly Core
+* WildFly
+* JBoss Log Manager
+
+=== Other Interested Projects
+
+* Any consumer of WildFly Core that uses the logging subsystem.
+
+== Requirements
+
+=== Hard Requirements
+
+* The attribute, `allow-log-routing`, is a simple boolean attribute which determines if the new or old behavior will be used.
+* When the new behavior is disabled the behavior should be the same as it currently works.
+    ** This means that when disabled the bug in https://issues.redhat.com/browse/WFCORE-3376[WFCORE-3376] would
+       still be present.
+* The `LogContext` should be associated with a deployments class loader.
+    ** This means that if the TCCL is set to the deployment class loader log messages will be routed to the deployments
+       log context.
+    ** That may include messages that previously went to the `server.log` could end up in a handler defined on the
+       deployments log context. This is desired behavior IMO.
+* Loggers created on any log context should be routable to another log context based on the TCCL, if enabled.
+
+=== Nice-to-Have Requirements
```

Comment:
What about the migration? This RFE changes the behaviour somehow, so we should have a plan about the migration between EAP7 -> EAP 8. Migration will be implemented as a part of the following 2 RFEs, but I believe that this analysis should cover at least some recommended way for a migration of logging configuration

* https://issues.redhat.com/browse/EAP7-1704
* https://issues.redhat.com/browse/EAP7-1778

Reply:
I do wonder if we should default to the current behavior instead of defaulting to the new behavior like I suggested. That would make migration not really an issue I think.

Reply:
New EAP major version is the best version where we can change wrong behaviour to correct behaviour. Do you have some specific reasons for keeping the wrong behaviour as default?

Reply:
I'm not sure I'd call it the "wrong behavior" really, but the reason I think we might want to keep the current default is just that it won't create any migration issues. However, I've got no real string opinion either way.

Reply:
> I'm not sure I'd call it the "wrong behavior" really

Hm, ok, so maybe that we should finish the clarification of this analysis, there are still some open threads. Btw:

https://github.com/wildfly/wildfly-proposals/pull/281#discussion_r695546316:

> This is my current understanding of this example: ... this classes are already configured to use LogContext from custom profile, which is wrong

https://github.com/wildfly/wildfly-proposals/pull/281#discussion_r696001988:

> Yes. Your understanding is correct.

Reply:
Analysis still talks about https://issues.redhat.com/browse/WFCORE-3376 "bug". https://issues.redhat.com/browse/WFCORE-3376 itself is a jira with bug type. And the behaviour where logs from deployment A uses LogContext from deployment B really seems like a bug. On the other hand, "it won't create any migration issues" this doesn't looks like strong reasoning in my PoV. If "no migration issues" would be a requirement, we can't do any improvements.

Reply:
Well the bug cannot be fixed without this. However, this is more than just a bug fix because it will change the behavior of other logs which may be unexpected.

I guess what I meant by "won't create any migration issues" was just if the attribute is set to `false` there is no change in behavior.

Reply:
> if the attribute is set to false there is no change in behavior.

+1. So the migration plan is just to set the attribute to false. What about the way how to achieve this? Manually? Or by transformers? Or by some another way? cc @emmartins

Reply:
@marekkopecky If we default the attribute to `false` then there should be no migration issues. Only users that manually set the attribute to `true` will see the affects of this RFE.

Reply:
> Can you please explicitly mention in the analysis what will be the default value of `allow-log-routing` attribute? I see two resolved threads about this topic, but I still don't see this info mentioned in the analysis:
> 
> * [[WFCORE-4807] Add the analysis document for log routing with an optio… #281 (comment)](https://github.com/wildfly/wildfly-proposals/pull/281#discussion_r695561754)
> * [[WFCORE-4807] Add the analysis document for log routing with an optio… #281 (comment)](https://github.com/wildfly/wildfly-proposals/pull/281#discussion_r745565981)

Sorry about that @marekkopecky. I meant to do that when I resolved those comments and got side-tracked. It's fixed now.

Reply:
> If we default the attribute to false then there should be no migration issues

Ah, from what I understand from your last comment, default value would be true and migration path would be "set false". Anyway, as we discussed earlier, there is some bug, that is fixed by this change. And default behaviour should be without a bug in my PoV.

>  jamezp 10 days ago
> Well the bug cannot be fixed without this

Reply:
It's kind of a weird one I think. While it does fix a bug, it's not a common bug. It changes behavior more than just fixes a bug which is I thought opting for `false` seemed correct. I've got no strong opinion either way, but defaulting to `true` may be a surprise to users using per-deployment logging.

Reply:
Is there some option how to fix that bug without changing the behaviour in other cases? AFAIU new behaviour seems to be better, right? New EAP8 major release is the best opportunity when we can fix/change wrong/misleading behaviour

Reply:
@marekkopecky There is no way to fix the bug without this change. I'm personally fine if we default to the new behavior. If everyone agrees I'm happy to use `true` for the default. And you're right that EAP 8 is the release to do this in.